File: src/content/docs/topics/cs/coding-problems/linked-list/138-copy-list-with-random-pointer-approach2.ts
function assert(condition: boolean, msg: string = ''): void {
    if (!condition) throw new Error(msg || 'Assertion failed');
}

class Node {
    val: number;
    next: Node | null;
    random: Node | null;
    constructor(val: number = 0, next: Node | null = null, random: Node | null = null) {
        this.val = val; this.next = next; this.random = random;
    }
}

function copyRandomList(head: Node | null): Node | null {
    if (!head) return null;
    const oldToNew = new Map<Node, Node>();
    let cur: Node | null = head;
    while (cur) { oldToNew.set(cur, new Node(cur.val)); cur = cur.next; }
    cur = head;
    while (cur) {
        oldToNew.get(cur)!.next = cur.next ? oldToNew.get(cur.next)! : null;
        oldToNew.get(cur)!.random = cur.random ? oldToNew.get(cur.random)! : null;
        cur = cur.next;
    }
    return oldToNew.get(head)!;
}

assert(copyRandomList(null) === null);
const n1 = new Node(1); n1.random = n1;
const copy1 = copyRandomList(n1);
assert(copy1 !== null && copy1 !== n1 && copy1!.val === 1 && copy1!.random === copy1);
const a = new Node(7); const b = new Node(13); a.next = b; b.random = a;
const copy2 = copyRandomList(a);
assert(copy2 !== null && copy2 !== a && copy2!.val === 7 && copy2!.next!.val === 13 && copy2!.next!.random === copy2);
console.log('all tests pass');

File: src/content/docs/topics/cs/coding-problems/linked-list/138-copy-list-with-random-pointer-approach2.go
package main

import "fmt"

type Node struct {
	Val    int
	Next   *Node
	Random *Node
}

func assert(condition bool, msgs ...string) {
	if !condition {
		msg := "assertion failed"
		if len(msgs) > 0 {
			msg = msgs[0]
		}
		panic(msg)
	}
}

func copyRandomList(head *Node) *Node {
	if head == nil {
		return nil
	}
	oldToNew := make(map[*Node]*Node)
	cur := head
	for cur != nil {
		oldToNew[cur] = &Node{Val: cur.Val}
		cur = cur.Next
	}
	cur = head
	for cur != nil {
		if cur.Next != nil {
			oldToNew[cur].Next = oldToNew[cur.Next]
		}
		if cur.Random != nil {
			oldToNew[cur].Random = oldToNew[cur.Random]
		}
		cur = cur.Next
	}
	return oldToNew[head]
}

func runTests() {
	assert(copyRandomList(nil) == nil)

	n1 := &Node{Val: 1}
	n1.Random = n1
	copy1 := copyRandomList(n1)
	assert(copy1 != nil && copy1 != n1)
	assert(copy1.Val == 1)
	assert(copy1.Random == copy1)

	a := &Node{Val: 7}
	b := &Node{Val: 13}
	a.Next = b
	b.Random = a
	copy2 := copyRandomList(a)
	assert(copy2 != nil && copy2 != a)
	assert(copy2.Val == 7)
	assert(copy2.Next.Val == 13)
	assert(copy2.Next.Random == copy2)

	fmt.Println("all tests pass")
}

func main() { runTests() }
